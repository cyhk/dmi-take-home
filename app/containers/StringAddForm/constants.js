export const ADD_STRING = 'ADD_STRING';
export const ADD_STRING_SUCCESS = 'ADD_STRING_SUCCESS';
export const ADD_STRING_FAILURE = 'ADD_STRING_FAILURE';
