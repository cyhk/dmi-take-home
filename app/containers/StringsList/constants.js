export const GET_STRINGS = 'GET_STRINGS';
export const GET_STRINGS_SUCCESS = 'GET_STRINGS_SUCCESS';
export const GET_STRINGS_FAILURE = 'GET_STRINGS_FAILURE';
