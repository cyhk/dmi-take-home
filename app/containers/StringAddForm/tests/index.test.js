/**
 * Test the StringAddForm
 */

import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { browserHistory } from 'react-router-dom';

import { StringAddForm, mapDispatchToProps } from '../index';
import { ADD_STRING } from '../constants';
import configureStore from '../../../configureStore';

describe('<StringAddForm />', () => {
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <StringAddForm updating={false} updated={false} error={null} />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  // Because the component was written as a functional component,
  // it is difficult to unit test handleChange and handleSubmit.
  // I have written integration tests to test the functionality of
  // these two functions. React-testing-library also emphasises
  // integration tests, so it would be more fitting if we follow
  // their testing philosophy
  it('should update the input field when it is changed', () => {
    const component = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <StringAddForm updating={false} updated={false} error={null} />
        </IntlProvider>
      </Provider>,
    );

    const input = component.getByLabelText('New string:');
    fireEvent.change(input, { target: { value: 'testing!' } });
    expect(input.value).toBe('testing!');
  });

  it('should reset the input field when the submit button is clicked', () => {
    const component = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <StringAddForm updating={false} updated={false} error={null} />
        </IntlProvider>
      </Provider>,
    );

    const input = component.getByLabelText('New string:');
    const button = component.getByText('Add');

    fireEvent.click(button);

    expect(input.value).toBe('');
  });

  describe('mapDispatchToProps', () => {
    describe('addString', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.addString).toBeDefined();
      });

      it('should dispatch ADD_STRING when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const string = 'test';
        result.addString(string);
        expect(dispatch).toHaveBeenCalledWith({ type: ADD_STRING, string });
      });
    });
  });
});
