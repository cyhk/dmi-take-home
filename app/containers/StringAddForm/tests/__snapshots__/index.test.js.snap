// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StringAddForm /> should render and match the snapshot 1`] = `
<form
  class="Wrapper-sc-60yovz-0 gOUjES"
>
  <div>
    <label
      class="Label-vify0n-0 GSYfe"
      for="string"
    >
      New string:
    </label>
    <input
      class="Input-sc-19txxuq-0 gGqJwk"
      id="string"
      name="String"
      type="text"
      value=""
    />
  </div>
  <button
    class="Button-sc-1fkoyq4-0 fmLlsE"
    type="submit"
  >
    Add
  </button>
</form>
`;
